### PR TITLE
Feature/teams

### PR DESCRIPTION
**Pull request by eman-ijaz7vals**

### CODE REVIEW

**Comment by eman-ijaz7vals** [comment]
Cycle 0 -reviewed

File: app/controllers/teams_controller.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
+class TeamsController < ApplicationController
+  load_and_authorize_resource
+
+  def index
+    @all_teams = Team.all
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: should be loaded by cancan

File: app/controllers/teams_controller.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  def new
+    @team = Team.new
+  end
+
+  def show
+    @team = Team.find(params[:id])
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: cancan will automatically find team

File: app/controllers/teams_controller.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      ## member who created team is by default made as admin
+
+      redirect_to teams_path, team_created: t('.Team created Successfully')
+    else
+      render 'new', team_created: t('.Team not created')
+    end
+  end
+
+  def edit
+    @team = Team.find(params[:id])
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: load using cancan

File: app/controllers/teams_controller.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  def update
+    @team = Team.find(params[:id])
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: load using cancan

File: app/controllers/teams_controller.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    if @team.update(team_params)
+      redirect_to teams_path, team_updated: t('.Team Successfully updated!')
+    else
+      render 'edit'
+    end
+  end
+
+  def destroy
+    @team = Team.find(params[:id])
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: load using cancan

File: app/controllers/teams_controller.rb [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  def show
+    @team = Team.find(params[:id])
+    @team_memberships = TeamMembership.where('team_id = ? ', @team.id)
+    already_present_member_ids = @team_memberships.pluck(:user_id)
+    @all_members = User.where.not(id: already_present_member_ids).select(:id, :first_name)
+  end
+
+  def create
+    @team = Team.new(team_params)
+    @team.company_id = 1 ## need to be changed
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: it will be set automatically.. no need to write this line

File: app/controllers/teams_controller.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+    if @team.save
+      @user_id = current_user.id
+      @team_id = Team.last.id
+      TeamMembership.create!(is_team_admin: true, is_approved: false, team_id: @team_id, user_id: @user_id) ## person who
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: create team using current user like this:
current_user.teams.create!(options)

Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: is_approved should be true for admin..
and is_approved should be false for new member who request to join the team, until admin of team approves his request.

File: app/controllers/teams_controller.rb [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+  def new
+    @team = Team.new
+  end
+
+  def show
+    @team = Team.find(params[:id])
+    @team_memberships = TeamMembership.where('team_id = ? ', @team.id)
+    already_present_member_ids = @team_memberships.pluck(:user_id)
+    @all_members = User.where.not(id: already_present_member_ids).select(:id, :first_name)
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: optional: there is performance issue with this approach. try to use autocomplete gem or try to implement custom logic using jquery-ui autocomplete widget to implement searching

File: app/controllers/teams_controller.rb [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  # change status of user "is_approved = true" on request to join team
+  def join_team
+    @is_approved = true
+    add_membership(@is_approved)
+  end
+
+  def add_membership(is_approved)
+    @user_id = params[:user_id]
+    @team_id = params[:team_id]
+    @is_admin = if params[:join_admin][:result] == '0'
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: use a different name instead of 'result'

File: app/controllers/teams_controller.rb [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    params.require(:team).permit(:name, :company_id)
+  end
+
+  def team_membership_params
+    params.require(:team_memberships).permit(:is_team_admin, :is_approved, :team_id, :user_id)
+  end
+
+  ## add member in team membership
+  def add_member
+    @is_approved = false
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: should be true

File: app/controllers/teams_controller.rb [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    add_membership(@is_approved)
+  end
+
+  # change status of user "is_approved = true" on request to join team
+  def join_team
+    @is_approved = true
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: should be false

File: app/controllers/teams_controller.rb [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  def team_membership_params
+    params.require(:team_memberships).permit(:is_team_admin, :is_approved, :team_id, :user_id)
+  end
+
+  ## add member in team membership
+  def add_member
+    @is_approved = false
+    add_membership(@is_approved)
```
Comment (abubakarazeem) [inline app/controllers/teams_controller.rb]: only update the is_approved field, no need to create membership again..

File: app/views/teams/index.html.erb [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+              <% if can? :manage, Team %> 
+                    <%= link_to 'Manage Team', team_path(team),class: 'btn btn-sm btn-block btn-outline-info' %><!--  shown to admins only -->
+              <% else %>
+                 
+                    <%= link_to 'View Team', team_path(team),class: 'btn btn-sm btn-block btn-outline-info' %>
+                  
+
+
+                    <hr>
+                    <% if TeamMembership.where('team_id = ? and user_id= ? and is_approved= ?', team.id, current_user.id, true).present? %>  <!-- change User.last.id to current user id ## CANCANCAN-->
```
Comment (abubakarazeem) [inline app/views/teams/index.html.erb]: don't write queries in views..

File: app/views/teams/show.html.erb [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+          <button type="submit" class="btn btn-primary btn-sm">Add member</button>
+        </div>
+      </div>   
+
+    </div>
+    <% end %>
+<% end %>
+  <!-- end of authorization -->
+
+  <div style="height:450px; padding-top: 50px" class="overflow-auto">
```
Comment (abubakarazeem) [inline app/views/teams/show.html.erb]: optional: define a class in css and use that class

File: spec/controllers/teams_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,200 @@
+require 'rails_helper'
+
+RSpec.describe TeamsController, type: :controller do
```
Comment (abubakarazeem) [inline spec/controllers/teams_controller_spec.rb]: make sure tests are working.. after reverse merge

File: app/controllers/teams_controller.rb (hunk 8 above)
Comment (eman-ijaz7vals) [reply app/controllers/teams_controller.rb]: Will see this performance things later!

File: app/controllers/teams_controller.rb [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+  load_and_authorize_resource
+  # GET /index
+  def index
+    respond_to do |format|
+      format.html
+    end
+  end
+  # GET /new
+  def new
+    @team = Team.new
```
Comment (Anmol79) [inline app/controllers/teams_controller.rb]: no need

File: app/controllers/application_controller.rb [hunk 17]
```diff
@@ -1,6 +1,8 @@
 class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
+
+  # protect_from_forgery with: :exception
```
Comment (Anmol79) [inline app/controllers/application_controller.rb]: remove

File: app/controllers/teams_controller.rb [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+    respond_to do |format|
+      format.html
+    end
+  end
+  # POST /create
+  def create
+    if @team.save
+      team_created = true
+      flash[:notice] = t('team.create.success')
+      begin
```
Comment (Anmol79) [inline app/controllers/teams_controller.rb]: move to model and should be in transaction

File: app/helpers/teams_helper.rb [hunk 19]
```diff
@@ -0,0 +1,5 @@
+module TeamsHelper
+  def requested(team_id, user_id)
+    TeamMembership.find_by('team_id = ? and user_id= ? and is_approved= ?', team_id, user_id, false).present?
```
Comment (Anmol79) [inline app/helpers/teams_helper.rb]: use current_company

File: app/models/team.rb [hunk 20]
```diff
@@ -1,4 +1,6 @@
 class Team < ActiveRecord::Base
+	sequenceid :company , :teams
```
Comment (Anmol79) [inline app/models/team.rb]: spaces

**Pull request merged**